### ISSUE TITLE
Review and/or fix changes to the VAE architecture for v4.1

### ISSUE THREAD

**Issue by Kevin-Chen0** [Photrek/Coupled-VAE#7]
I have looked at the VAE architecture from Keras [here](https://keras.io/examples/generative/vae/), and propose to make the following changes to VAE v.4.1:

1. Make the encoder/decoder as tfk.Model instead of tfk.Layer
2. Have the encoder/decoder pair passed into the VAE model like other hyperparameters
3. Remove the VAE wrapper class and move the necessary functions into the VAEModel class, including:
      -`train_step()`
      -`compute_loss()`
      -`log_normal_pdf()`
      -`generate_and_save_images()`
4. Use `.fit()` to train VAEModel instead of `.train()`. 

The benefits of this approach is simplication of the code by not having use a wrapper class. This also aligns with how Keras model usually trains the model by using `.fit()`. 

If this is the way to go, I have created a new VAE version [vae_v4.1a_wip.ipynb](https://github.com/Photrek/Machine-Learning-for-Complex-Systems/blob/master/vae/vae_v4_1a_wip.ipynb), but has encounter a handful of errors:
 
1. Encoder/Decoder model "has not been built" and therefore I am not able to view the 'summary()'. Alternatively, we may need to write in the way that Keras tutorial has already provided but I prefer not to.

```
latent_dim = 2

encoder_inputs = keras.Input(shape=(28, 28, 1))
x = layers.Conv2D(32, 3, activation="relu", strides=2, padding="same")(encoder_inputs)
x = layers.Conv2D(64, 3, activation="relu", strides=2, padding="same")(x)
x = layers.Flatten()(x)
x = layers.Dense(16, activation="relu")(x)
z_mean = layers.Dense(latent_dim, name="z_mean")(x)
z_log_var = layers.Dense(latent_dim, name="z_log_var")(x)
z = Sampling()([z_mean, z_log_var])
encoder = keras.Model(encoder_inputs, [z_mean, z_log_var, z], name="encoder")
encoder.summary()
```

2. Passing in `tensorflow_datasets` object types into VAEModel to fit. For some reason, in the Keras tutorial, they had to concatenate `x_train` and `x_test` into `mnist_digits` and then fit just `mnist_digits`. They can do that because `x_train` and `x_test` are numpy arrays. In our case, we currently pass in `fm_train_dataset` and `fm_train_dataset` separately. Do we have to mimic what keras is doing here. Would `tensorflow.python.data.ops.dataset_ops.ShuffleDataset` even work here?

```
(x_train, _), (x_test, _) = keras.datasets.mnist.load_data()
mnist_digits = np.concatenate([x_train, x_test], axis=0)
mnist_digits = np.expand_dims(mnist_digits, -1).astype("float32") / 255

vae = VAE(encoder, decoder)
vae.compile(optimizer=keras.optimizers.Adam())
vae.fit(mnist_digits, epochs=30, batch_size=128)
```

[Here](https://www.tensorflow.org/api_docs/python/tf/keras/Model) is the link to keras Model class.

3. Ability for it to train without the `.train()` from the wrapper class. Do we simply move those contents into the `.train_step()` function in VAEModel? How about the `test_step()` or `fit()` function? Do we need to override them in from tfk.Model?

4.  Where to put the `generate_and_save_images()`? Put that into the `.train_step()` so that there can be plots for every train step? Also, `generate_and_save_images()` inputs only test images while `self.compute_loss()` inputs only train images. So it may be problematic if we have the concatenate train and test images together as mentioned in 2. 

Can you take a look at the following?

Thanks!
